numpy/_core/src/umath/_scaled_float_dtype.c

```diff
@@ -1040,27 +1014,19 @@ sfloat_init_ufuncs(void) {
     }
 
     /*
-     * Add a promoter for both directions of multiply with double.
+     * Add a dtype-based promoter for multiply, matching whenever one
+     * of the operands is a scaled float.
      */
     int res = -1;
-    PyArray_DTypeMeta *double_DType = &PyArray_DoubleDType;
-
-    PyArray_DTypeMeta *promoter_dtypes[3] = {
-            &PyArray_SFloatDType, double_DType, NULL};
-
     PyObject *promoter = PyCapsule_New(
             &promote_to_sfloat, "numpy._ufunc_promoter", NULL);
     if (promoter == NULL) {
         return -1;
     }
-    res = sfloat_add_loop("multiply", promoter_dtypes, promoter);
-    if (res < 0) {
-        Py_DECREF(promoter);
-        return -1;
-    }
-    promoter_dtypes[0] = double_DType;
-    promoter_dtypes[1] = &PyArray_SFloatDType;
-    res = sfloat_add_loop("multiply", promoter_dtypes, promoter);
+
+    PyObject *ufunc = sfloat_get_ufunc("multiply");
+    res = PyUFunc_AddDTypeBasedPromoter(
+        (PyUFuncObject *)ufunc, &PyArray_SFloatDType, promoter);
```

Comment:
This is probably dangerous (this matches even for non-floats), so this is really just a test, happy to remove.